[android] Handle mediacodec callback with a handler thread

mingchou opened the pull request:
[android] Handle mediacodec callback with a handler thread

Now the MediaCodec callbacks and the activity lifecycle callbacks
are called on the main thread. When pausing YouTube, the MediaCodec
callbacks may be blocked until the lifecycle callbacks are done.
This may cause frame drops on a resource-limited device when the
MediaCodec callback to notify a newly decoded frame is blocked and
we run out of decoded frames. Running the MediaCodec callbacks on a
handler thread could avoid this problem.

To ensure the callbacks won't be called when the media decoder is
destroyed, clear the native media codec bridge handle in java layer
when destroying MediaDecoder.

b/316008643

borongc commented on starboard/android/shared/player_components_factory.h:
```diff
@@ -423,7 +447,11 @@ class PlayerComponentsFactory : public starboard::shared::starboard::player::
 
       using starboard::shared::starboard::media::AudioStreamInfo;
-      auto decoder_creator = [enable_flush_during_seek](
+      bool audio_use_mediacodec_callback_thread =
```
As dropping frames is likely to happen on video decoder, instead of audio decoder, do you think we can only move video mediacodec callback with a handler thread?

If yes, would you be able to verify if we only move video mediacodec callback with a handler thread can solve the issue?

borongc commented on starboard/android/shared/media_decoder.cc:
```diff
@@ -159,6 +162,9 @@ MediaDecoder::~MediaDecoder() {
   }
 
   if (is_valid()) {
+    // Clear the native media codec bridge handle in java layer to prevent
```
`media_codec_bridge_->Flush()` is after this `media_codec_bridge_->Stop()`, which makes Cobalt fails to access MediaCodec.

borongc commented on starboard/android/shared/media_decoder.cc:
```diff
@@ -159,6 +162,9 @@ MediaDecoder::~MediaDecoder() {
   if (is_valid()) {
+    // Clear the native media codec bridge handle in java layer to prevent
+    // calling mediacodec callbacks with an invalid handle.
+    media_codec_bridge_->Stop();
```
`media_codec_bridge_->Flush()` is after this `media_codec_bridge_->Stop()`, which makes Cobalt fails to access MediaCodec.

mingchou replied:
Revised.

borongc commented on starboard/android/apk/app/src/main/java/dev/cobalt/media/MediaCodecBridge.java:
```diff
@@ -827,6 +846,11 @@ public void release() {
       Log.e(TAG, "Cannot release media codec", e);
     }
     mMediaCodec = null;
+    if (mCallbackThread != null) {
```
Consider to add `synchronized (this) {}` block as the callbacks are guarded by `this` object.

mingchou replied:
Revised

mingchou committed:
[android] Handle mediacodec callback with a handler thread

Now the MediaCodec callbacks and the activity lifecycle callbacks
are called on the main thread. When pausing YouTube, the MediaCodec
callbacks may be blocked until the lifecycle callbacks are done.
This may cause frame drops on a resource-limited device when the
MediaCodec callback to notify a newly decoded frame is blocked and
we run out of decoded frames. Running the MediaCodec callbacks on a
handler thread could avoid this problem.

Change-Id: I0f5582103c85402e3d8355a6d82a48d244a98fae

borongc commented on starboard/android/shared/player_components_factory.h:
```diff
@@ -80,6 +80,11 @@ constexpr bool kForceFlushDecoderDuringReset = false;
+// Use a new created handler thread to handle mediacodec callbacks. This
+// prevents the mediacodec callbacks are blocked by activity lifecycle
+// callbacks.
+constexpr bool kUseHandlerThread = false;
```
Consider to rename this to `kUseMediaCodecCallbackThread` to align with the `mediacodeccallbackthread`.

mingchou replied:
revised

borongc commented on starboard/android/shared/media_codec_bridge.h:
```diff
@@ -174,6 +174,7 @@ class MediaCodecBridge {
       int tunnel_mode_audio_session_id,
       bool force_big_endian_hdr_metadata,
       int max_video_input_size,
+      bool use_callback_thread,
```
Align with name `use_mediacodec_callback_thread`.

mingchou replied:
revised

mingchou committed:
[android] Clear native media codec bridge when destroying MediaDecoder

Clear the native media codec bridge handle in java layer to prevent calling
mediacodec callbacks with an invalid handle when destroying MediaDecoder.

Change-Id: I3d6cbfebfff027f65063dd4af760db624561b593

borongc commented on starboard/android/shared/player_components_factory.h:
```diff
@@ -271,10 +276,22 @@ class PlayerComponentsFactory : public starboard::shared::starboard::player::
+        MimeType video_mime_type(creation_parameters.video_mime());
+        if (!video_mime_type.is_valid() ||
+            !video_mime_type.ValidateBoolParameter(
+                "mediacodeccallbackthread")) {
```
As some browsers don't recognize MIME attributes, we should re-shape this feature to use H5VCC flags instead of mime type params, to avoid additional infrastructure issues.

We are working on the PRs of https://github.com/youtube/cobalt/pull/3186 and https://github.com/youtube/cobalt/pull/2830, and will provide a general way to set the flag via H5VCC.

As a result, this PR will be reshaped from using MIME attribute to H5VCC flags after the above PRs are merged.

borongc commented:
The PR (https://github.com/youtube/cobalt/pull/3740) might fix the main issue mentioned in this PR.
